### PR TITLE
Sparse symmetric eri

### PR DESCRIPTION
**Pull request by AlexanderMath**
`N=500` sto3g 100 carbon atoms in 10x10 grid **(this directly computes `J-K/2*HYB_B3LYP` no full dft)**

image

<hr>
**Results below are for full DFT computation.**  `N=80` sto3g 16 carbon atoms in 4x4 grid. 

image

<hr>

`N=180` sto3g 36 carbon atoms in 6x6 grid

image

image

<hr>

`N=360` sto3g 64 carbons atoms in 8x8 grid (goes OOM due to bad tilin

image

image

### CODE REVIEW

File: nanoDFT/sparse_symmetric_ERI.py [hunk 1]
```diff
@@ -62,35 +80,153 @@ def num_repetitions_fast(value):
...
+
+    print("reshape to 16 IPUs pmap", time.time()-start)
+    nonzero_indices = nonzero_indices.reshape(nipu, batches, -1)
+    nonzero_distinct_ERI = nonzero_distinct_ERI.reshape(nipu, batches, -1)
+    print(nonzero_indices.shape)
+    print(nonzero_distinct_ERI.shape)
+
+    print("call pmap", time.time()-start)
+    print("[%i]"%mol.nao_nr())
+    diff_JK = jax.pmap( sparse_symmetric_einsum, in_axes=(0,0,None,None,None, None), static_broadcasted_argnums=(5,), backend=backend, axis_name="p") (nonzero_distinct_ERI, nonzero_indices, dm, mask, diff_JK, args.backend)
```
Comment (mihaipgc) [inline nanoDFT/sparse_symmetric_ERI.py:222]: This line throws an error, I think `diff_JK` can be removed from the arguments, i.e.:

`diff_JK = jax.pmap( sparse_symmetric_einsum, in_axes=(0,0,None,None,None), static_broadcasted_argnums=(4,), backend=backend, axis_name="p") (nonzero_distinct_ERI, nonzero_indices, dm, mask, args.backend) `

File: nanoDFT/sparse_symmetric_ERI.py [hunk 2]
```diff
@@ -62,35 +80,153 @@ def num_repetitions_fast(value):
...
+    nonzero_distinct_ERI = distinct_ERI[nonzero_indices].astype(np.float32)
+    if not args.skip: 
+        print("dense: ", dense_ERI.shape, dense_ERI.nbytes/10**6)
+    print("sparse: ", distinct_ERI.shape, distinct_ERI.nbytes/10**6)
+    print("sparse+symmetric: ", nonzero_distinct_ERI.shape, nonzero_distinct_ERI.nbytes/10**6)
+    print("per ipu", nonzero_distinct_ERI.nbytes/10**6/16*2) # obs: most time might be spent compiling loading ERI from H->device! 
+    print("----------\n")
+
+    # write code that profiles get_indices/take/segment_sum independently. 
+    if args.prof:
```
Comment (mihaipgc) [inline nanoDFT/sparse_symmetric_ERI.py:182]: Perhaps remove this part as it looks deprecated since some of the functions and variables no longer exist (and/or create an issue to address this?)

File: nanoDFT/experimental_pmap_nanoDFT.py [hunk 3]
```diff
@@ -215,28 +218,56 @@ def nanoDFT(mol, opts):
...
-        # ERI = tensors[6]  # ERI is just the 6th element in tensors
-        sparse_ERI = sparse_representation(ERI)
-        remainder = sparse_ERI[-1].shape[0] % 16
-        if remainder != 0:
-            sparse_ERI = [np.pad(a, (0, -remainder+16)) for a in sparse_ERI]
-        sparse_ERI = [a.reshape(16, -1) for a in sparse_ERI]
-        # tensors[6] = sparse_ERI
+        in_axes=(None, None, None, None, None, None, None, None, None, None, None, (None, None, None), 0, 0, [0, 0]))
+
+        if False:
```
Comment (mihaipgc) [inline nanoDFT/experimental_pmap_nanoDFT.py:223]: Unreachable block, should we remove it or have a flag to control it?

File: nanoDFT/compute_indices.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+ij_pair get_i_j(int32_t val) {
+    ij_pair result;
+    result.i = max_val(val);
+    result.j = val - result.i * (result.i + 1) / 2;
+    return result;
+}
+
+
+class indices : public Vertex {
```
Comment (mihaipgc) [inline nanoDFT/compute_indices.cpp:43]: naming suggestion: SymmetryIndices

File: nanoDFT/experimental_pmap_nanoDFT.py [hunk 5]
```diff
@@ -167,16 +170,16 @@ def init_dft_tensors_cpu(mol, opts, DIIS_iters=9):
     L_inv           = np.linalg.inv(np.linalg.cholesky(O))          # (N,N)
     #if opts.backend != "ipu": ERI = mol.intor("int2e_sph")          # (N,N,N,N)=(66,66,66,66) for C6H6.
     #else:                     ERI = None # will be computed on device
-    ERI = mol.intor("int2e_sph")
+    ERI = 0 #mol.intor("int2e_sph")
```
Comment (mihaipgc) [inline nanoDFT/experimental_pmap_nanoDFT.py:173]: Is this intended / temporary? If this is the case, then it might help to add a comment (e.g. "TODO fix later") to find occurrences easily later as several variables are zeroed out at the moment

**Review by mihaipgc — CHANGES_REQUESTED** [review]
One small issue at `line 222` in `sparse_symmetric_ERI.py`. I also added a few comments/questions which you are free to ignore if they're not necessary at this stage.

File: nanoDFT/sparse_symmetric_ERI.py [hunk 6]
```diff
@@ -62,35 +80,153 @@ def num_repetitions_fast(value):
...
+    N = mol.nao_nr()
+    print("[%i]"%mol.nao_nr())
+    if not args.skip: 
+        print("dense_ERI", time.time()-start)
+        dense_ERI = mol.intor("int2e_sph", aosym="s1")
+    print("distinct_ERI", time.time()-start)
+    distinct_ERI = mol.intor("int2e_sph", aosym="s8")
+    print("Minao init", time.time()-start)
+    dm = pyscf.scf.hf.init_guess_by_minao(mol)         
+    HYB_B3LYP = 0.2
```
[the remark this answers is not in the document]
Comment (AlexanderMath) [reply nanoDFT/sparse_symmetric_ERI.py]: Done.

File: nanoDFT/sparse_symmetric_ERI.py (hunk 2 above)
Comment (AlexanderMath) [reply nanoDFT/sparse_symmetric_ERI.py:181]: Adding instead `TODO: fix broken profile code, and add profile individually for np.take and segment_sum, potentially with custom poplar implementations`

File: nanoDFT/experimental_pmap_nanoDFT.py (hunk 3 above)
Comment (AlexanderMath) [reply nanoDFT/experimental_pmap_nanoDFT.py:223]: Removing.

File: nanoDFT/compute_indices.cpp (hunk 4 above)
Comment (AlexanderMath) [reply nanoDFT/compute_indices.cpp]: Done.

File: nanoDFT/experimental_pmap_nanoDFT.py (hunk 5 above)
Comment (AlexanderMath) [reply nanoDFT/experimental_pmap_nanoDFT.py]: Added TODO.

**Review by mihaipgc — APPROVED** [review]
LGTM
